stan/math/rev/core/vari.hpp

```diff
@@ -252,14 +274,32 @@ class vari_value<T, std::enable_if_t<is_eigen_dense_base<T>::value>>
    * @tparam S A dense Eigen type that is convertible to `value_type`
    * @param x Value of the constructed variable.
    */
-  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  template <typename S, require_convertible_t<S&, T_val>* = nullptr>
   explicit vari_value(S&& x)
-      : val_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
-            x.size())),
-        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
-            x.size())),
-        adj_(make_adj(x)),
-        val_(make_val(x)) {
+      : Base(),
+        adj_(plain_type_t<T_adj>::Zero(x.rows(), x.cols())),
+        val_(std::forward<S>(x)) {
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+  /**
+   * Construct a dense Eigen variable implementation from a value and adjoint.
+   *
+   * All constructed variables are added to the stack. Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @tparam R A dense Eigen type that is convertible to `T_val`
+   * @tparam S A dense Eigen type that is convertible to `T_adj`
+   * @param val Value of the constructed variable.
+   * @param adj Adjoint of the constructed variable.
+   */
+  template <typename R, typename S, require_convertible_t<R&, T_val>* = nullptr,
+            require_convertible_t<S&, T_adj>* = nullptr>
+  vari_value(R&& val, S&& adj)
+      : Base(), adj_(std::forward<S>(adj)), val_(std::forward<R>(val)) {
     ChainableStack::instance_->var_stack_.push_back(this);
   }
```

Comment:
Should this be private/protected? Or do you think we will want to set the adjoints directly outside of the class?

Reply:
I'm not sure. I think there might be some other use cases to construct vari with pre-allocated memory.